vmas/simulator/environment/environment.py

```diff
@@ -490,41 +513,40 @@ def _set_action(self, action, agent):
             if not self.multidiscrete_actions:
                 # This bit of code translates the discrete action (taken from a space that
                 # is the cartesian product of all action spaces) into a multi discrete action.
-                # For example, if agent.action_size=4, it will mean that the agent will have
-                # 4 actions each with 3 possibilities (stay, decrement, increment).
-                # The env will have a space Discrete(3**4).
-                # This code will translate the action (with shape [n_envs,1] and range [0,3**4)) to an
-                # action with shape [n_envs,4] and range [0,3).
-                n_actions = self.get_agent_action_space(agent).n
-                action_range = torch.arange(n_actions, device=self.device).expand(
-                    self.world.batch_dim, n_actions
-                )
-                physical_action = action
-                action_range = torch.where(action_range == physical_action, 1.0, 0.0)
-                action_range = action_range.view(
-                    (self.world.batch_dim,)
-                    + (3,) * agent.action_size
-                    + (self.world.dim_c,)
-                    * (1 if not agent.silent and self.world.dim_c != 0 else 0)
+                # This is done by iteratively taking the modulo of the action and dividing by the
+                # number of actions in the current action space, which treats the action as if
+                # it was the "flat index" of the multi-discrete actions. E.g. if we have
+                # action spaces [2, 3, 4], action 0 corresponds to the actions [0, 0, 0],
+                # action 1 corresponds to the action [1, 0, 0], action 2 corresponds
+                # to the action [0, 1, 0], action 3 corresponds to the action [1, 1, 0], etc.
+                flat_action = action.squeeze(-1)
+                actions = []
+                nvec = list(agent.action_nvec) + (
+                    [self.world.dim_c]
+                    if not agent.silent and self.world.dim_c != 0
+                    else []
                 )
-                action = action_range.nonzero()[:, 1:]
+                for n in nvec:
+                    actions.append(flat_action % n)
+                    flat_action = flat_action // n
+                action = torch.stack(actions, dim=-1)
```

Comment:
This changes the logic: before we had 0-> [0,0,0], 1-> [0,0,1], 2-> [0,0,2], 3-> [0,1,0], 4-> [0,1,1] and so on